mimic/apache.go

```diff
@@ -0,0 +1,321 @@
+	path string
+}
+
+const timeFormat = "Mon, 02 Jan 2006 15:04:05 GMT"
+
+var (
+	Host         string
+	Port         string
+	lastModified = time.Now().Format(timeFormat)
+	etag         = makeETag()
+)
+
+// MimicApache mimics the behaviour of an unconfigured Apache web server 2.4.7
+// (the one installed by 'apt-get install apache2') running on Ubuntu 14.04.
+// Set 'Host' and 'Port' before calling it.
+func MimicApache(w http.ResponseWriter, req *http.Request) {
+	conn, _, err := w.(http.Hijacker).Hijack()
+	if err != nil {
+		panic("fail to hijack, should not happen")
+	}
+	path := req.URL.Path
+	// remove extra leading slash
+	if len(path) > 0 && path[0] == '/' {
+		i := 1
+		for ; i < len(path) && path[i] == '/'; i++ {
+		}
+		if i > 0 {
+			i--
+		}
+		path = path[i:]
+	}
+	m := apacheMimic{conn, req, path}
+	if req.Host == "" {
+		m.writeError(badRequestHeader, badRequestBody)
+		return
+	}
+	switch req.Method {
+	case "CONNECT":
+		m.writeError(badRequestHeader, badRequestBody)
+	case "GET", "POST":
+		switch path {
+		case "/", "/index.html":
+			m.ok(indexHeader, indexDotHTML)
+		case "/icons/ubuntu-logo.png":
+			m.ok(logoHeader, ubuntuLogo)
+		default:
+			m.writeError(notFoundHeader, notFoundBody)
+		}
+	case "HEAD":
+		switch path {
+		case "/", "/index.html":
+			m.ok(indexHeader, nil)
+		case "/icons/ubuntu-logo.png":
+			m.ok(logoHeader, nil)
+		default:
+			m.writeError(notFoundHeaderWhenHead, nil)
+		}
+	case "OPTIONS":
+		switch path {
+		case "/", "/index.html":
+			m.ok(optionsHeader, nil)
+		case "/icons/ubuntu-logo.png":
+			m.ok(optionsHeader, nil)
+		default:
+			m.writeError(optionsHeaderWhenNotFound, nil)
+		}
+	default:
+		if m.path == "/" {
+			m.path = "/index.html"
+		}
+		switch {
+		case !KNOWN_METHODS[req.Method]:
+			m.writeError(notImplementedHeader, notImplementedBody)
+		case !ALLOWED_METHODS[req.Method]:
+			m.writeError(methodNotAllowedHeader, methodNotAllowedBody)
+		case KNOWN_URIS[path]:
+			//m.ok()
+		default:
+			m.writeError(notFoundHeader, notFoundBody)
+		}
+	}
+}
+
+func (f *apacheMimic) ok(header *template.Template, body []byte) {
+	var buf bytes.Buffer
+	err := header.Execute(&buf, f.collectVars())
+	if err != nil {
+		panic(fmt.Sprintf("execute template err: %s", err))
+	}
+	// ignore any errors writing back to connection
+	_, _ = buf.WriteTo(f.conn)
+	_, _ = f.conn.Write(body)
+	_ = f.conn.Close()
+}
+
+func (f *apacheMimic) writeError(header, body *template.Template) {
+	var headBuf bytes.Buffer
+	var bodyBuf bytes.Buffer
+	vars := f.collectVars()
+	if body != nil {
+		err := body.Execute(&bodyBuf, vars)
+		if err != nil {
+			panic(fmt.Sprintf("execute template err: %s", err))
+		}
+		vars.ContentLength = bodyBuf.Len()
+	}
+	err := header.Execute(&headBuf, vars)
+	if err != nil {
+		panic(fmt.Sprintf("execute template err: %s", err))
+	}
+	// ignore any errors writing back to connection
+	_, _ = headBuf.WriteTo(f.conn)
+	_, _ = bodyBuf.WriteTo(f.conn)
+	_ = f.conn.Close()
```

Comment:
Good catch! I just wrote it here and didn't notice the potential leak.

On Friday, October 30, 2015, Ulysses Aalto <notifications@github.com> wrote:

> In mimic/apache.go
> <https://github.com/getlantern/chained-server/pull/27#discussion_r43491862>
> :
>
> > +	vars := f.collectVars()
> > +	if body != nil {
> > +		err := body.Execute(&bodyBuf, vars)
> > +		if err != nil {
> > +			panic(fmt.Sprintf("execute template err: %s", err))
> > +		}
> > +		vars.ContentLength = bodyBuf.Len()
> > +	}
> > +	err := header.Execute(&headBuf, vars)
> > +	if err != nil {
> > +		panic(fmt.Sprintf("execute template err: %s", err))
> > +	}
> > +	// ignore any errors writing back to connection
> > +	_, _ = headBuf.WriteTo(f.conn)
> > +	_, _ = bodyBuf.WriteTo(f.conn)
> > +	_ = f.conn.Close()
>
> I think you put this here because deferring is more costly, is that right?
> I don't think is a problem, because we are panicking on errors, but since
> those panics happen in goroutines the program might not terminate and leave
> this connections open. That said, if the goroutine panics the templates are
> buggy and need a server fix anyway.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/getlantern/chained-server/pull/27/files#r43491862>.
>